adding an advanced method for utilising cache with fetch (#7765)
cloudflare/cloudflare-docs: content/workers/examples/cache-using-fetch.md

This option is a version of the `cacheTtl` feature which chooses a TTL based on the response's status code and does not automatically set `cacheEverything: true`. If the response to this request has a status code that matches, Cloudflare will cache for the instructed time, and override cache directives sent by the origin. You can review [details on the `cacheTtl` feature on the Request page](/workers/runtime-apis/request/#requestinitcfproperties).

{{</content-column>}}

{{<content-column>}}

## Cache based on file extensions and using arrays to pass through variables

> You can combine the functionalities mentioned above into one script and use variables instead so that it can be repurposed for other use cases. In this case, it can be done for a media streaming platform.

Reviewer: ```suggestion
You can combine the functionalities mentioned above into one Worker script and use variables instead so that it can be repurposed for other use cases. In the below example, this can be done for a media streaming platform.
```
Is the last sentence saying that the example below is for a media streaming platform? If that is so, changed the sentence to be more descriptive of the below content.

{{</content-column>}}

{{<tabs labels="js/esm | js/sw">}}
{{<tab label="js/esm" default="true">}}

> ```js

Reviewer: ```suggestion
```js
---
filename: 
---
```
Filename needed.

export default {
  async fetch(request) {

  // Instantiate new URL to make it mutable
  const newRequest = new URL(request.url)

>   // Set const to be used in the array later on

Reviewer: ```suggestion
  // Set `const` to be used in the array later on
```

  const cacheAssets = [
      {asset: 'video', key: customCacheKey, regex: /(.*\/Video)|(.*\.(m4s|mp4|ts|avi|mpeg|mpg|mkv|bin|webm|vob|flv|m2ts|mts|3gp|m4v|wmv|qt))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'image', key: queryCacheKey, regex: /(.*\/Images)|(.*\.(jpg|jpeg|png|bmp|pict|tif|tiff|webp|gif|heif|exif|bat|bpg|ppm|pgn|pbm|pnm))/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'frontEnd', key: queryCacheKey, regex: /^.*\.(css|js)/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'audio', key: customCacheKey, regex: /(.*\/Audio)|(.*\.(flac|aac|mp3|alac|aiff|wav|ogg|aiff|opus|ape|wma|3gp))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'directPlay', key: customCacheKey, regex: /.*(\/Download)/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'manifest', key: customCacheKey, regex: /^.*\.(m3u8|mpd)/, info: 0, ok: 3, redirects: 2, clientError: 1, serverError: 0 }
  ]

>   // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find we'll be using the regex to match on file extensions for caching

Reviewer: ```suggestion
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find you will be using the `regex` to match on file extensions for caching
```
`regex` object?

  const { asset, regex, ...cache } = cacheAssets.find( ({regex}) => newRequest.pathname.match(regex)) ?? {}

  const newResponse = await fetch(request,
          { cf:
              {
                  cacheKey: cache.key,
                  polish: false,
                  cacheEverything: true,
                  cacheTtlByStatus: {
                      '100-199': cache.info,
                      '200-299': cache.ok,
                      '300-399': cache.redirects,
                      '400-499': cache.clientError,
                      '500-599': cache.serverError
                      },
                  cacheTags: [ 
                      'static'
                      ]
              },
          
          })

  const response = new Response(newResponse.body, newResponse)

  // For debugging purposes
  response.headers.set('debug', JSON.stringify(cache))
  return response
  }
}
```
{{</tab>}}
{{<tab label="js/sw">}}

> ```js

Reviewer: Filename needed

Author: Ah, what do you mean by filename here?

addEventListener('fetch', event => {
    return event.respondWith(handleRequest(event.request));
})

async function handleRequest(request) {

// Instantiate new URL to make it mutable
const newRequest = new URL(request.url)

// Set const to be used in the array later on
const customCacheKey = `${newRequest.hostname}${newRequest.pathname}`
const queryCacheKey = `${newRequest.hostname}${newRequest.pathname}${newRequest.search}`

> // Here we set all variables needed to manipulate Cloudflare's cache using the fetch API in the cf object, we'll be passing these variables in the objects down

Reviewer: ```suggestion
// Here we set all variables needed to manipulate Cloudflare's cache using the fetch API in the cf object, you will be passing these variables in the objects down
```
Same comment as above

const cacheAssets = [
    {asset: 'video', key: customCacheKey, regex: /(.*\/Video)|(.*\.(m4s|mp4|ts|avi|mpeg|mpg|mkv|bin|webm|vob|flv|m2ts|mts|3gp|m4v|wmv|qt))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'image', key: queryCacheKey, regex: /(.*\/Images)|(.*\.(jpg|jpeg|png|bmp|pict|tif|tiff|webp|gif|heif|exif|bat|bpg|ppm|pgn|pbm|pnm))/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'frontEnd', key: queryCacheKey, regex: /^.*\.(css|js)/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'audio', key: customCacheKey, regex: /(.*\/Audio)|(.*\.(flac|aac|mp3|alac|aiff|wav|ogg|aiff|opus|ape|wma|3gp))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'directPlay', key: customCacheKey, regex: /.*(\/Download)/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'manifest', key: customCacheKey, regex: /^.*\.(m3u8|mpd)/, info: 0, ok: 3, redirects: 2, clientError: 1, serverError: 0 }
]

> // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find we'll be using the regex to match on file extensions for caching

Reviewer: ```suggestion
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find use the `regex` to match on file extensions for caching
```
same comment as above

This option is a version of the `cacheTtl` feature which chooses a TTL based on the response's status code and does not automatically set `cacheEverything: true`. If the response to this request has a status code that matches, Cloudflare will cache for the instructed time, and override cache directives sent by the origin. You can review [details on the `cacheTtl` feature on the Request page](/workers/runtime-apis/request/#requestinitcfproperties).

{{</content-column>}}

{{<content-column>}}

## Cache based on file extensions and using arrays to pass through variables

You can combine the functionalities mentioned above into one Worker script and use variables instead so that it can be repurposed for other use cases. In the below example, this can be done for a media streaming platform.

{{</content-column>}}

{{<tabs labels="js/esm | js/sw">}}
{{<tab label="js/esm" default="true">}}

```js
---

> filename: 

Reviewer: ```suggestion
filename: index.js
```

---
export default {
  async fetch(request) {

  // Instantiate new URL to make it mutable
  const newRequest = new URL(request.url)

  // Set `const` to be used in the array later on
  const customCacheKey = `${newRequest.hostname}${newRequest.pathname}`
  const queryCacheKey = `${newRequest.hostname}${newRequest.pathname}${newRequest.search}`

>   // Here we set all variables needed to manipulate Cloudflare's cache using the fetch API in the cf object, you will be passing these variables in the objects down

Reviewer: ```suggestion
  // Below, you set all variables needed to manipulate Cloudflare's cache using the fetch API in the cf object, you will be passing these variables in the objects down
```

Reviewer: Is `cf object` referring to an object in the code? Unsure of what this is

Author: It's this: https://developers.cloudflare.com/workers/runtime-apis/request/#requestinitcfproperties, which is intiated in the fetch at the end, where I passed in all the TTL values

  const cacheAssets = [
      {asset: 'video', key: customCacheKey, regex: /(.*\/Video)|(.*\.(m4s|mp4|ts|avi|mpeg|mpg|mkv|bin|webm|vob|flv|m2ts|mts|3gp|m4v|wmv|qt))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'image', key: queryCacheKey, regex: /(.*\/Images)|(.*\.(jpg|jpeg|png|bmp|pict|tif|tiff|webp|gif|heif|exif|bat|bpg|ppm|pgn|pbm|pnm))/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'frontEnd', key: queryCacheKey, regex: /^.*\.(css|js)/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'audio', key: customCacheKey, regex: /(.*\/Audio)|(.*\.(flac|aac|mp3|alac|aiff|wav|ogg|aiff|opus|ape|wma|3gp))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'directPlay', key: customCacheKey, regex: /.*(\/Download)/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'manifest', key: customCacheKey, regex: /^.*\.(m3u8|mpd)/, info: 0, ok: 3, redirects: 2, clientError: 1, serverError: 0 }
  ]

>   // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find you will be using the `regex` to match on file extensions for caching

Reviewer: 1) What connection does the MDN link have to the code below? This needs a short explanation.
2) "the `regex`..." object? field? This seems like it's missing something

Author: For 1 and 2: You can say: the .find method is used to find elements in an array (cacheAssets), in this case regex, which can passed to the .match method to match on file extensions to cache, since they are many media types in the array. So if a person wants to add more types, they can just update the array instead of messing around with the logic.

  const { asset, regex, ...cache } = cacheAssets.find( ({regex}) => newRequest.pathname.match(regex)) ?? {}

  const newResponse = await fetch(request,
          { cf:
              {
                  cacheKey: cache.key,
                  polish: false,
                  cacheEverything: true,
                  cacheTtlByStatus: {
                      '100-199': cache.info,
                      '200-299': cache.ok,
                      '300-399': cache.redirects,
                      '400-499': cache.clientError,
                      '500-599': cache.serverError
                      },
                  cacheTags: [ 
                      'static'
                      ]
              },
          
          })

  const response = new Response(newResponse.body, newResponse)

  // For debugging purposes
  response.headers.set('debug', JSON.stringify(cache))
  return response
  }
}
```
{{</tab>}}
{{<tab label="js/sw">}}

> ```js

Reviewer: ```suggestion
```js
---
filename: index.js
---
```

const cacheAssets = [
    {asset: 'video', key: customCacheKey, regex: /(.*\/Video)|(.*\.(m4s|mp4|ts|avi|mpeg|mpg|mkv|bin|webm|vob|flv|m2ts|mts|3gp|m4v|wmv|qt))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'image', key: queryCacheKey, regex: /(.*\/Images)|(.*\.(jpg|jpeg|png|bmp|pict|tif|tiff|webp|gif|heif|exif|bat|bpg|ppm|pgn|pbm|pnm))/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'frontEnd', key: queryCacheKey, regex: /^.*\.(css|js)/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'audio', key: customCacheKey, regex: /(.*\/Audio)|(.*\.(flac|aac|mp3|alac|aiff|wav|ogg|aiff|opus|ape|wma|3gp))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'directPlay', key: customCacheKey, regex: /.*(\/Download)/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'manifest', key: customCacheKey, regex: /^.*\.(m3u8|mpd)/, info: 0, ok: 3, redirects: 2, clientError: 1, serverError: 0 }
]

> // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find use the `regex` to match on file extensions for caching

Reviewer: Same comment as above.

{{</content-column>}}

{{<content-column>}}

## Customize cache behavior based on request file type

Using custom cache keys and overrides based on response code, you can write a Worker that sets the TTL based on the response status code from origin, and request file type.

The following example demonstrates how you might use this to cache requests for streaming media assets:

{{</content-column>}}

{{<tabs labels="js/esm | js/sw">}}
{{<tab label="js/esm" default="true">}}

```js
---
header: index.js
---
export default {
  async fetch(request) {

  // Instantiate new URL to make it mutable
  const newRequest = new URL(request.url)

  const customCacheKey = `${newRequest.hostname}${newRequest.pathname}`
  const queryCacheKey = `${newRequest.hostname}${newRequest.pathname}${newRequest.search}`

> 

Reviewer: ```suggestion

// Different asset types usually have different caching strategies. Most of the time media content such as audio, videos and images that are not user-generated content would not need to be updated often so a long TTL would be best. However, with HLS streaming, manifest files usually are set with short TTLs so that playback will not be affected, as this files contain the data that the player would need. By setting each caching strategy for categories of asset types in an object within an array, you can solve complex needs when it comes to media content for your application

```

This option is a version of the `cacheTtl` feature which chooses a TTL based on the response's status code and does not automatically set `cacheEverything: true`. If the response to this request has a status code that matches, Cloudflare will cache for the instructed time, and override cache directives sent by the origin. You can review [details on the `cacheTtl` feature on the Request page](/workers/runtime-apis/request/#requestinitcfproperties).

{{</content-column>}}

{{<content-column>}}

> ## Cache based on file extensions and using arrays to pass through variables

Reviewer 2: ```suggestion
## Customize cache behavior based on request file type
```

> You can combine the functionalities mentioned above into one Worker script and use variables instead so that it can be repurposed for other use cases. In the below example, this can be done for a media streaming platform.

Reviewer 2: ```suggestion
Using custom cache keys and overrides based on response code, you can write a Worker that sets the TTL based on the response status code from origin, and request file type.

The following example demonstrates how you might use this to cache requests for streaming media assets:
```

{{</content-column>}}

{{<tabs labels="js/esm | js/sw">}}
{{<tab label="js/esm" default="true">}}

```js
---
header: index.js
---
export default {
  async fetch(request) {

>   // Instantiate new URL to make it mutable

Reviewer 2: ```suggestion 
```
Don't see this mutated anywhere?

Author: I just do that as a practice if down the line I would want to modify the code or mutate the URL/request

  const newRequest = new URL(request.url)

>   // Set `const` to be used in the array later on

Reviewer 2: ```suggestion

```

Think this is self explanatory enough

  const customCacheKey = `${newRequest.hostname}${newRequest.pathname}`
  const queryCacheKey = `${newRequest.hostname}${newRequest.pathname}${newRequest.search}`

>   // Here we set all variables needed to manipulate Cloudflare's cache using the fetch API in the cf object, you will be passing these variables in the objects down

Reviewer: ```suggestion
  // Below, set all variables needed to manipulate Cloudflare's cache using the fetch API in the `cf` object. You will be passing these variables in the objects down below.
```

>   // Set all variables needed to manipulate Cloudflare's cache using the fetch API in the `cf` object. You will be passing these variables in the objects down below.

Reviewer 2: ```suggestion
```

  const cacheAssets = [
      {asset: 'video', key: customCacheKey, regex: /(.*\/Video)|(.*\.(m4s|mp4|ts|avi|mpeg|mpg|mkv|bin|webm|vob|flv|m2ts|mts|3gp|m4v|wmv|qt))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },

>       {asset: 'image', key: queryCacheKey, regex: /(.*\/Images)|(.*\.(jpg|jpeg|png|bmp|pict|tif|tiff|webp|gif|heif|exif|bat|bpg|ppm|pgn|pbm|pnm))/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },

Reviewer 2: Why does this use `queryCacheKey`, while other examples use `customCacheKey`?

Author: It's an example to show that parts of an application can be using query params to vary height, size, tags, quality etc

      {asset: 'frontEnd', key: queryCacheKey, regex: /^.*\.(css|js)/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'audio', key: customCacheKey, regex: /(.*\/Audio)|(.*\.(flac|aac|mp3|alac|aiff|wav|ogg|aiff|opus|ape|wma|3gp))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'directPlay', key: customCacheKey, regex: /.*(\/Download)/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
      {asset: 'manifest', key: customCacheKey, regex: /^.*\.(m3u8|mpd)/, info: 0, ok: 3, redirects: 2, clientError: 1, serverError: 0 }

>   ]

Reviewer 2: Think it would be good to explain somewhere why one might want to define different behavior for different asset types. What problem does this solve? etc.

Reviewer: @erfianugrah , can you add a suggestion to what line you think this snippet would be appropriate in?

Author: You can put on line 202

>   // the `.find` method is used to find elements in an array (`cacheAssets`), in this case, `regex`, which can passed to the .`match` method to match on file extensions to cache, since they are many media types in the array. If you want to add more types, update the array. Refer to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find for more information.

Reviewer 2: ```suggestion
```

                  polish: false,
                  cacheEverything: true,
                  cacheTtlByStatus: {
                      '100-199': cache.info,
                      '200-299': cache.ok,
                      '300-399': cache.redirects,
                      '400-499': cache.clientError,
                      '500-599': cache.serverError
                      },
                  cacheTags: [ 
                      'static'
                      ]
              },
          
          })

  const response = new Response(newResponse.body, newResponse)

  // For debugging purposes
  response.headers.set('debug', JSON.stringify(cache))
  return response
  }
}
```
{{</tab>}}
{{<tab label="js/sw">}}

```js
---
header: index.js
---
addEventListener('fetch', event => {
    return event.respondWith(handleRequest(event.request));
})

async function handleRequest(request) {

// Instantiate new URL to make it mutable
const newRequest = new URL(request.url)

// Set `const` to be used in the array later on
const customCacheKey = `${newRequest.hostname}${newRequest.pathname}`
const queryCacheKey = `${newRequest.hostname}${newRequest.pathname}${newRequest.search}`

> // Below, you set all variables needed to manipulate Cloudflare's cache using the fetch API in the `cf` object. You will be passing these variables in the objects down.

Reviewer: ```suggestion
// Below, you set all variables needed to manipulate Cloudflare's cache using the fetch API in the `cf` object. You will be passing these variables in the objects down below.
```

const cacheAssets = [
    {asset: 'video', key: customCacheKey, regex: /(.*\/Video)|(.*\.(m4s|mp4|ts|avi|mpeg|mpg|mkv|bin|webm|vob|flv|m2ts|mts|3gp|m4v|wmv|qt))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'image', key: queryCacheKey, regex: /(.*\/Images)|(.*\.(jpg|jpeg|png|bmp|pict|tif|tiff|webp|gif|heif|exif|bat|bpg|ppm|pgn|pbm|pnm))/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'frontEnd', key: queryCacheKey, regex: /^.*\.(css|js)/, info: 0, ok: 3600, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'audio', key: customCacheKey, regex: /(.*\/Audio)|(.*\.(flac|aac|mp3|alac|aiff|wav|ogg|aiff|opus|ape|wma|3gp))/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'directPlay', key: customCacheKey, regex: /.*(\/Download)/, info: 0, ok: 31556952, redirects: 30, clientError: 10, serverError: 0 },
    {asset: 'manifest', key: customCacheKey, regex: /^.*\.(m3u8|mpd)/, info: 0, ok: 3, redirects: 2, clientError: 1, serverError: 0 }
]

> // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find use the `regex` to match on file extensions for caching

Reviewer: ```suggestion
// the `.find` method is used to find elements in an array (`cacheAssets`), in this case, `regex`, which can passed to the .`match` method to match on file extensions to cache, since they are many media types in the array. If you want to add more types, update the array. Refer to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find for more information.
```

const { asset, regex, ...cache } = cacheAssets.find( ({regex}) => newRequest.pathname.match(regex)) ?? {}

const newResponse = await fetch(request,
        { cf:
            {
                cacheKey: cache.key,
                polish: false,
                cacheEverything: true,
                cacheTtlByStatus: {
                    '100-199': cache.info,
                    '200-299': cache.ok,
                    '300-399': cache.redirects,
                    '400-499': cache.clientError,
                    '500-599': cache.serverError
                    },
                cacheTags: [ 
                    'static'
                    ]
            },
        
        })

const response = new Response(newResponse.body, newResponse)

// For debugging purposes
response.headers.set('debug', JSON.stringify(cache))
return response
}
```
{{</tab>}}

> {{</tabs>}}

Reviewer 2: Feel free to omit adding a service worker example, and just have an ESM example.

{{</content-column>}}

{{<content-column>}}

## Customize cache behavior based on request file type

Using custom cache keys and overrides based on response code, you can write a Worker that sets the TTL based on the response status code from origin, and request file type.

The following example demonstrates how you might use this to cache requests for streaming media assets:

{{</content-column>}}

{{<tabs labels="js/esm | js/sw">}}
{{<tab label="js/esm" default="true">}}

```js
---
header: index.js
---
export default {
  async fetch(request) {

  // Instantiate new URL to make it mutable
  const newRequest = new URL(request.url)

  const customCacheKey = `${newRequest.hostname}${newRequest.pathname}`
  const queryCacheKey = `${newRequest.hostname}${newRequest.pathname}${newRequest.search}`

> // Different asset types usually have different caching strategies. Most of the time media content such as audio, videos and images that are not user-generated content would not need to be updated often so a long TTL would be best. However, with HLS streaming, manifest files usually are set with short TTLs so that playback will not be affected, as this files contain the data that the player would need. By setting each caching strategy for categories of asset types in an object within an array, you can solve complex needs when it comes to media content for your application

Reviewer: ```suggestion
  // Different asset types usually have different caching strategies. Most of the time media content such as audio, videos and images that are not user-generated content would not need to be updated often so a long TTL would be best. However, with HLS streaming, manifest files usually are set with short TTLs so that playback will not be affected, as this files contain the data that the player would need. By setting each caching strategy for categories of asset types in an object within an array, you can solve complex needs when it comes to media content for your application
```

> // Different asset types usually have different caching strategies. Most of the time media content such as audio, videos and images that are not user-generated content would not need to be updated often so a long TTL would be best. However, with HLS streaming, manifest files usually are set with short TTLs so that playback will not be affected, as this files contain the data that the player would need. By setting each caching strategy for categories of asset types in an object within an array, you can solve complex needs when it comes to media content for your application

Reviewer: ```suggestion
  // Different asset types usually have different caching strategies. Most of the time media content such as audio, videos and images that are not user-generated content would not need to be updated often so a long TTL would be best. However, with HLS streaming, manifest files usually are set with short TTLs so that playback will not be affected, as this files contain the data that the player would need. By setting each caching strategy for categories of asset types in an object within an array, you can solve complex needs when it comes to media content for your application
```
